### PR TITLE
KV health check for DynamoDB

### PR DESCRIPTION
**Pull request by idanovo**
### Linked Issue

Closes #[(Insert issue number closed by this PR)](https://github.com/treeverse/lakeFS/issues/5294)

### CODE REVIEW

File: docs/reference/configuration.md [hunk 1]
```diff
@@ -60,9 +60,10 @@ This reference uses `.` to denote the nesting of values.
     + `database.dynamodb.aws_region` `(string : )` - AWS Region of database instance
     + `database.dynamodb.aws_profile` `(string : )` - AWS named profile to use
     + `database.dynamodb.aws_access_key_id` `(string : )` - AWS access key ID
-    + `database.dynamodb.aws_secret_access_key` `(string : )` - AWS secret access key
+    + `database.dynamodb.aws_secret_access_key` `(string : )` - AWS secret access key.
```
Comment (idanovo) [inline docs/reference/configuration.md:63]:
```suggestion
    + `database.dynamodb.aws_secret_access_key` `(string : )` - AWS secret access key
```

File: pkg/config/config.go [hunk 2]
```diff
@@ -141,6 +141,11 @@ type Config struct {
 			AwsProfile         string `mapstructure:"aws_profile"`
 			AwsAccessKeyID     string `mapstructure:"aws_access_key_id"`
 			AwsSecretAccessKey string `mapstructure:"aws_secret_access_key"`
+
+			// Interval to run health check for the DynamoDB instance
```
Comment (nopcoder) [inline pkg/config/config.go:145]: Code doc should start with the name of the func/variable - `HealthCheckInterval` in this case.
No need to specify the default - it is done in the code later in this file.

File: pkg/config/defaults.go [hunk 3]
```diff
@@ -97,6 +97,7 @@ func setDefaults(local bool) {
 
 	viper.SetDefault("database.dynamodb.table_name", "kvstore")
 	viper.SetDefault("database.dynamodb.scan_limit", 1024)
+	viper.SetDefault("database.dynamodb.health_check_interval", 0*time.Minute)
```
Comment (nopcoder) [inline pkg/config/defaults.go:100]: no need - as far as I know the default for Duration type is zero.

File: pkg/kv/dynamodb/store.go [hunk 4]
```diff
@@ -94,18 +96,22 @@ func (d *Driver) Open(ctx context.Context, kvParams kvparams.Config) (kv.Store,
 		return nil, fmt.Errorf("%w: %s", kv.ErrSetupFailed, err)
 	}
 
-	return &Store{
+	s := &Store{
 		svc:    svc,
-		params: params,
-	}, nil
+		params: params}
```
Comment (nopcoder) [inline pkg/kv/dynamodb/store.go:101]:
```suggestion
		params: params,
	}
```

File: pkg/kv/dynamodb/store.go [hunk 5]
```diff
@@ -94,18 +96,22 @@ func (d *Driver) Open(ctx context.Context, kvParams kvparams.Config) (kv.Store,
...
+	s.StartPeriodicCheck()
+	return s, nil
 }
 
 // isTableExist will try to describeTable and return bool status, error is returned only in case err != ResourceNotFoundException
 func isTableExist(ctx context.Context, svc *dynamodb.DynamoDB, table string) (bool, error) {
 	_, err := svc.DescribeTableWithContext(ctx, &dynamodb.DescribeTableInput{
 		TableName: aws.String(table),
 	})
+	const operation = "isTableExist"
```
Comment (nopcoder) [inline pkg/kv/dynamodb/store.go:112]: The operation is `DescribeTable`

File: pkg/kv/dynamodb/store.go [hunk 6]
```diff
@@ -94,18 +96,22 @@ func (d *Driver) Open(ctx context.Context, kvParams kvparams.Config) (kv.Store,
...
 	if err != nil {
+		dynamoFailures.WithLabelValues(operation).Inc()
 		if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == dynamodb.ErrCodeResourceNotFoundException {
 			return false, nil
 		}
```
Comment (nopcoder) [inline pkg/kv/dynamodb/store.go:117]: Verify that for common setup - non-exists table is not consider as an error we like to report as an error.

File: pkg/kv/params/database.go [hunk 7]
```diff
@@ -53,4 +53,9 @@ type DynamoDB struct {
 	AwsProfile         string
 	AwsAccessKeyID     string
 	AwsSecretAccessKey string
+
+	// Interval to run health check for the DynamoDB instance
+	// Won't run when is equal or less than 0
+	// Default: 0
```
Comment (nopcoder) [inline pkg/kv/params/database.go:59]: 1. code doc format
2. think it is maybe too much documentation - keep it here or on the configuration (here is fine)
3. remove the default - it is based on the configuration.

File: pkg/kv/dynamodb/store.go [hunk 8]
```diff
@@ -455,3 +463,43 @@ func (e *EntriesIterator) Err() error {
...
+		logging.Default().Debug("DynamoDB health check passed!")
+	} else {
+		logging.Default().WithError(err).Debug("DynamoDB health check failed")
+	}
+}
+
+func (s *Store) StopPeriodicCheck() {
+	if s.ticker != nil {
+		logging.Default().Info("Stopping DynamoDB health check")
+		s.ticker.Stop()
```
Comment (nopcoder) [inline pkg/kv/dynamodb/store.go:503]: 1. Logging - we should create logger in the driver level which holds the name of the store type - check local adapter. When possible we should pass logger or create one relevant to the service that log information.
2. Need wait group to wait under the check goroutine is done. This is just to be nice with the caller.
3. If we signal the goroutine to stop using a channel - just close the channel and move the ticker into `StartPeriodicCheck` (not a member) + defer close, so when we close the channel in `StopPeriodcCheck` the ticker will be stopped.

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/kv/dynamodb/store.go [hunk 9]
```diff
@@ -455,3 +469,46 @@ func (e *EntriesIterator) Err() error {
...
+	} else {
+		log.WithError(err).Debug("DynamoDB health check failed")
+	}
+}
+
+func (s *Store) StopPeriodicCheck() {
+	if s.cancel != nil {
+		s.logger.Info("Stopping DynamoDB health check")
+	}
+	s.wg.Wait()
```
Comment (itaiad200) [inline pkg/kv/dynamodb/store.go:513]: Will never return if `interval <= 0 `

Comment (itaiad200) [reply pkg/kv/dynamodb/store.go:513]: or panic..

File: pkg/kv/dynamodb/store.go [hunk 10]
```diff
@@ -455,3 +469,48 @@ func (e *EntriesIterator) Err() error {
...
+	s.cancel = make(chan bool)
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		defer func() { s.cancel = nil }()
+		s.logger.WithField("interval", interval).Debug("Starting DynamoDB health check")
+		// check first and loop for checking every interval
+		s.Check()
+		ticker := time.NewTicker(interval)
+		defer ticker.Stop()
```
Comment (nopcoder) [inline pkg/kv/dynamodb/store.go:488]: 1. Use single defer func() { ... }()
2. Don't make and use the channel from places, you may write to a nil channel. It is safe to always create a channel and call close(ch) on close. (like the fact that we use the waitgroup even when we do not start a goroutine)

File: pkg/kv/dynamodb/store.go [hunk 11]
```diff
@@ -455,3 +469,48 @@ func (e *EntriesIterator) Err() error {
...
+		log.WithError(err).Debug("DynamoDB health check failed")
+	}
+}
+
+func (s *Store) StopPeriodicCheck() {
+	if s.cancel != nil {
+		s.logger.Info("Stopping DynamoDB health check")
+		s.cancel <- true
+		s.wg.Wait()
+	}
```
Comment (nopcoder) [inline pkg/kv/dynamodb/store.go:515]: To simplify - just close the channel and wait for the wg.
Log probably in debug level - not so interesting.

**Review by nopcoder — CHANGES_REQUESTED** [review]

File: pkg/kv/params/database.go [hunk 12]
```diff
@@ -53,4 +53,6 @@ type DynamoDB struct {
 	AwsProfile         string
 	AwsAccessKeyID     string
 	AwsSecretAccessKey string
+
+	HealthCheckInterval time.Duration
```
Comment (nopcoder) [inline pkg/kv/params/database.go:57]: remove whitespace

File: pkg/kv/dynamodb/store.go [hunk 13]
```diff
@@ -455,3 +470,45 @@ func (e *EntriesIterator) Err() error {
...
+	success, err := isTableExist(context.Background(), s.svc, s.params.TableName)
+	if success {
+		log.Debug("DynamoDB health check passed!")
+	} else {
+		log.WithError(err).Debug("DynamoDB health check failed")
+	}
+}
+
+func (s *Store) StopPeriodicCheck() {
+	if s.cancel != nil {
```
Comment (nopcoder) [inline pkg/kv/dynamodb/store.go:510]: remove the if

**Review by nopcoder — APPROVED** [review]
LGTM - just remove the extra `if` statement that is no longer required.

**Pull request merged**